### ISSUE TITLE
HCL: The argument "labels" is required, but no definition was found.

### ISSUE THREAD

**Issue by dpedu** [hashicorp/packer#12301]
#### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

#### Overview of the Issue

Packer raises an error:

```
Error: Missing required argument

  on myimage/build.pkr.hcl line 17:
  (source code not available)

The argument "labels" is required, but no definition was found.
```

But the documentation says this field is NOT required, here: https://developer.hashicorp.com/packer/docs/templates/hcl_templates/expressions#dynamic-blocks

Additionally - the documentation on that page is not sufficient in that I am unable to determine what this `labels` argument is actually used for or what values to put in there.

#### Reproduction Steps

I have a build block like this:

```
build {
  dynamic "source" {      # <-------- this is "line 17" in my error above
    for_each = {
      ubuntu = data.amazon-ami.ubuntu_2004_amd64_eks_122.id
      azl2   = data.amazon-ami.azl_2_amd64_eks_122.id
    }
    content {
      name  = source.key
      value = source.value
    }
  }

  provisioner "salt-masterless" {
    ...
  }

  post-processor "manifest" {
    ...
  }
}
```

I've tried guessing what I'm supposed to put for `labels`. It seems to need to go within the `dynamic "source" {` block.

I tried `labels = ["foo", "bar"]` to see if I would get a clearer error message, which told me: `Blocks of type "source" require 1 label(s).` This error was on the line of the `labels` argument. 

Ok, so I tried just one entry:

```
Error: Unknown source type

  on myimage/build.pkr.hcl line 5:
  (source code not available)

known builders: [null vsphere-iso qemu proxmox-iso linode tencentcloud-cvm
amazon-ebs vagrant proxmox-clone amazon-ebssurrogate profitbricks virtualbox-vm
vmware-iso azure-chroot oneandone proxmox amazon-instance amazon-ebsvolume
cloudstack jdcloud parallels-iso alicloud-ecs ncloud vsphere-clone lxd
digitalocean hyperone virtualbox-iso hyperv-vmcx googlecompute openstack
amazon-chroot hyperv-iso yandex file hcloud docker virtualbox-ovf ucloud-uhost
vmware-vmx lxc azure-arm triton parallels-pvm azure-dtl
```

This error was on the line the `source {` block opens. 

I'm building `amazon-ebs` AMIs, so maybe I'm supposed to use `amazon-ebs` as a label? It's in the output of supposedly valid ones printed above.

But no, that just prints the same error message.

Edit: this allow appears to break differentiation of output during parallel builds. My example build creates two AMIs in parallel but it labels them both `amazon-ebs.ami`. I can't pass in a second label for the second build, per the error above. Something here is broken.

### Packer version

From `packer version`

`Packer v1.8.6`

### Operating system and Environment details

MacOS 12.2 on Arm

### Log Fragments and crash.log files

N/A

**Comment by nywilken** [comment]
Hi @dpedu this is not supported at this time. It is a feature that is on our roadmap. I'm going to close this issue as it is a duplicate of https://github.com/hashicorp/packer/issues/12164. Please feel free to add any additional comments to the active thread.

**Comment by dpedu** [comment]
@nywilken can you reopen this? The issue you referenced, #12301 does not look similar. That issue deals with dynamic `provisioner` blocks, where as mine deals with dynamic `source` blocks. Dynamic source blocks are supported as I'm using them today.

**Comment by nywilken** [comment]
Hi @dpedu thanks for following up on this issue. I made a mistake in understanding the requests. The `labels` argument that is required here is the label of the defined source block `source "amzon-ebs" "example"` that is being referenced by the generated source blocks. 

Below is a working example of what I believe you are trying to achieve. 

```hcl

variable "sensitive" {
  default = "I am sensitive"
  sensitive = true
}

source "null" "first-example" {
}


build {
  name = "roles"

      dynamic "source" {
          for_each = {
            consul = "none",
            ubuntu = "none"
          }
          labels = ["null.first-example"]
          content {
             name = source.key
             communicator = source.value
          }
      }

  provisioner "shell-local" {
    inline = ["echo ${source.name} and ${source.type}, the sensitive var: ${var.sensitive}"]
  }
}
```

```
~>  packer build null-sensitive-variables.pkr.hcl
roles.null.consul: output will be in this color.
roles.null.ubuntu: output will be in this color.

==> roles.null.ubuntu: Running local shell script: /var/folders/jd/zhl6kpsn1156cmqdyttt22xm0000gp/T/packer-shell2825454605
==> roles.null.consul: Running local shell script: /var/folders/jd/zhl6kpsn1156cmqdyttt22xm0000gp/T/packer-shell1130209094
    roles.null.consul: consul and null, the sensitive var: <sensitive>
Build 'roles.null.consul' finished after 217 milliseconds 150 microseconds.
    roles.null.ubuntu: ubuntu and null, the sensitive var: <sensitive>
Build 'roles.null.ubuntu' finished after 275 milliseconds 761 microseconds.

==> Wait completed after 275 milliseconds 788 microseconds

==> Builds finished. The artifacts of successful builds are:
--> roles.null.consul: Did not export anything. This is the null builder
--> roles.null.ubuntu: Did not export anything. This is the null builder
```

**Comment by nywilken** [comment]
Closing as I believe this is resolved. If I am mistaken please let me know.
